### PR TITLE
feat(analysis): pre-flight size gates for dissolve, buffer, and layer masks

### LINKED ISSUE #694: feat(analysis): pre-flight per-operation feature_count gates (dissolve OOM guard, buffer output ceiling)
From the 2026-07-25 analysis scalability audit. Two failure modes converge on one mechanism: an at-enqueue, per-operation gate on `dataset.feature_count` (already loaded on the ORM row in the materialize route).

**Dissolve can OOM-kill the database container.** `ST_Union` as an aggregate accumulates every input geometry before the cascaded union runs (~1 GB practical wall), against `shared_buffers=512MB` and the 2 GB `mem_limit` on the db service. Memory grows faster than the 300s statement timeout elapses, so the kernel OOM-kills a backend and the postmaster resets **every connection in the cluster** — the whole app blips, all tenants. `ST_UnaryUnion(ST_Collect(...))` does not change the ceiling (identical accumulation). Grouped dissolve is worse under HashAggregate (all groups' arrays held simultaneously); GroupAggregate holds one at a time, and without stats on `by_field` the planner's choice is effectively random.
- Guard: reject above a `feature_count` threshold before the CTAS (`backend/app/processing/analysis/tasks.py`, pre-CTAS) or at enqueue with a clear 422.
- Plus `SET LOCAL enable_hashagg = off` for the grouped-dissolve CTAS to force the bounded sorted path.
- Ceiling-raiser later (L): hierarchical union (union per grid cell, then union the cells).

**Buffer is the only amplifying operation and has zero byte accounting.** Vector/PostGIS-resident datasets contribute 0 bytes to storage quota by design (`backend/app/modules/quota/service.py:34-42` defers to cloud metering), and `check_upload_quota` is called with `incoming_bytes=0` at enqueue. Buffer turns one API call that uploads nothing into an arbitrarily large table (500k inputs × ~32 output vertices at the 100 km distance ceiling), bounded only by the dataset-count quota, on shared infrastructure.
- Cap `feature_count` for buffer specifically (S), or estimate `feature_count × vertices-per-feature` against a configurable ceiling next to `MAX_BUFFER_METERS` (`backend/app/platform/analysis_sql.py:31`) (M).
- Belt-and-braces: post-CTAS `pg_total_relation_size` check before registration, failing into the existing cleanup path (which already drops the table correctly).

Caveat both gates honestly: `feature_count` is an ingest-time snapshot — a warning threshold, not a correctness gate.

### PR DESCRIPTION
**Pull request by ishiland**
## What

Batch 2 (first slice) from the 2026-07-25 analysis scalability audit: the blast-radius guards that keep one analysis request from taking down shared infrastructure.

- **Dissolve OOM guard (#694).** `ST_Union` accumulates every input geometry before the cascaded union runs; around ~1M polygons it OOM-kills the 2 GB db container and the postmaster resets **every connection in the cluster**. Dissolve is now rejected at enqueue above 250k features (422 with an actionable message), and the CTAS runs with `SET LOCAL enable_hashagg = off` so grouped dissolves aggregate one group at a time instead of holding all groups' union states simultaneously.
- **Buffer output ceiling (#694).** Buffer is the only output-amplifying operation and vector datasets carry no byte quota by design. Gated at 500k features at enqueue, plus a post-CTAS `pg_total_relation_size` backstop (2 GB) that fails the job before the rewrite/index/registration phases — the enqueue gates read the cached ingest-time `feature_count` snapshot, which can be stale; the backstop can't be.
- **Layer-mask cap (#693, first half).** A layer-sourced clip mask is unioned whole before any row limit can bite, inside the preview's 10-second budget; a few thousand realistic polygons already blow it. Capped at 1,000 features in the shared `_load_mask_dataset` loader, so preview and materialize agree. The per-row semi-join redesign for preview stays open in #693.

All caps read the cached catalog `feature_count` and are documented in-code as blast-radius guards, not exact gates. No API-schema changes (no OpenAPI/SDK regen needed).

Closes #694. Part of #693.

## Verification

- `pytest tests/test_analysis_materialize.py tests/test_analysis_preview.py tests/test_layering.py` — 99 passed (6 new tests: enqueue gates incl. boundary values, per-op selectivity, mask cap on both endpoints, hashagg SET LOCAL presence for dissolve / absence otherwise, size-backstop failure path).
- `make openapi-check`, `ruff check`, `ruff format --check` — clean.

https://claude.ai/code/session_01X1Wd3Frww9F2eovfMkUbG2

### CODE REVIEW

**Comment by ishiland** [comment]
@codex review

File: backend/app/processing/analysis/tasks.py [hunk 1]
```diff
@@ -351,8 +357,29 @@ async def _materialize(
...
+                # enqueue-time feature cap is the first line of defense;
+                # this is the second.
+                await session.execute(text("SET LOCAL enable_hashagg = off"))
             await session.execute(text(f"CREATE TABLE {out_ref} AS {select_sql}"))
             out_table_created = True
+            size_bytes = await session.scalar(
+                text(
+                    "SELECT pg_total_relation_size(CAST(:ref AS regclass))"
+                ).bindparams(ref=out_ref)
+            )
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/processing/analysis/tasks.py]: Confirmed and fixed in the commit above. The ceiling is now enforced twice via a shared `_enforce_output_size` helper: a cheap early exit right after the CTAS (a table already over the limit must not hold the single fleet-wide worker slot through the rewrite phases), and the authoritative check after `add_4326_column` — post-rewrite, so heap + TOAST + GIST all count — before the ANALYZE/commit. The spy test pins both probes and the second one's position between the 4326 rewrite and the ANALYZE.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/datasets/api/router_analysis.py [hunk 2]
```diff
@@ -147,6 +173,17 @@ async def analysis_materialize_endpoint(
     """
     dataset = await _load_vector_dataset(db, dataset_id, user)
 
+    max_features = _MAX_SOURCE_FEATURES.get(body.operation)
+    if max_features is not None and (dataset.feature_count or 0) > max_features:
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/datasets/api/router_analysis.py]: Confirmed and fixed in the commit above. Both gates now count via a new `resolve_source_feature_count(db, dataset, cap=...)` (domain layer, exported through the service façade): the cached snapshot when present, otherwise a live `SELECT count(*) FROM (SELECT 1 FROM t LIMIT cap+1)` against the physical table — bounded, so the probe can't itself become the expensive scan, and resolved with the tenant schema the same way the worker's non-sandbox reads are. Tests cover a NULL-snapshot dataset rejected by the live probe (cap patched below the true row count), the admit path above it, and the NULL-snapshot mask layer.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/datasets/api/router_analysis.py [hunk 3]
```diff
@@ -147,6 +179,19 @@ async def analysis_materialize_endpoint(
     """
     dataset = await _load_vector_dataset(db, dataset_id, user)
 
+    max_features = _MAX_SOURCE_FEATURES.get(body.operation)
+    if max_features is not None:
+        source_count = await resolve_source_feature_count(db, dataset, cap=max_features)
+        if source_count > max_features:
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/datasets/api/router_analysis.py:176]: Confirmed and fixed in the commit above. The ceilings now live in `app.platform.analysis_sql` (already the shared home of MAX_BUFFER_METERS, importable from both the catalog router and the processing worker without crossing layer boundaries), and `_materialize` re-runs the LIMIT-bounded live counts for the source and the layer mask immediately before building the SQL — failing the job with the same actionable message before any table is created. Tests cover both worker rechecks (source cap and mask cap patched below the live row counts).

**Comment by ishiland** [comment]
@codex review

File: backend/app/processing/analysis/tasks.py [hunk 4]
```diff
@@ -140,6 +151,74 @@ async def _fail_cancelled_job(
...
+    """Fail the build when the output relation is over MAX_OUTPUT_BYTES.
+
+    ``pg_total_relation_size`` counts heap, TOAST, and indexes, so the
+    post-rewrite call measures what the registered dataset will actually
+    occupy.
+    """
+    result = await session.execute(
+        text("SELECT pg_total_relation_size(CAST(:ref AS regclass))").bindparams(
+            ref=out_ref
+        )
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/processing/analysis/tasks.py]: Confirmed and fixed in the commit above. `_enforce_output_size` now takes (schema, table_name) and resolves the size via `pg_total_relation_size(c.oid)` from a `pg_class`/`pg_namespace` join using the recognized `:schema` bind — the same pattern as `_list_carry_columns` and the orphan-name probe — so the tenant hook sees the schema and the lookup no longer depends on regclass name resolution. (A string-literal `'…'::regclass` inline would not have worked either: the hook masks string literals before scanning.)

**Comment by ishiland** [comment]
@codex review
